harness/determined/pytorch/_trainer.py

```diff
@@ -0,0 +1,233 @@
+            steps_completed=steps_completed,
+            latest_checkpoint=latest_checkpoint,
+            local_training=self._local_training,
+            test_mode=test_mode,
+            reporting_period=reporting_period,
+            searcher_metric_name=searcher_metric_name,
+            checkpoint_policy=checkpoint_policy,
+            step_zero_validation=step_zero_validation,
+            max_length=max_length,
+            det_profiler=self._det_profiler,
+            global_batch_size=global_batch_size,
+        )
+
+        trial_controller.run()
+
+
+def _initialize_distributed_backend() -> Optional[core.DistributedContext]:
+    info = det.get_cluster_info()
+
+    distributed_backend = det._DistributedBackend()
+    if distributed_backend.use_horovod():
+        hvd.require_horovod_type("torch", "PyTorchTrial is in use.")
+        hvd.init()
+        return core.DistributedContext.from_horovod(horovod.hvd)
+    elif distributed_backend.use_torch():
+        if torch.cuda.is_available():
+            dist.init_process_group(backend="nccl")  # type: ignore
+        else:
+            dist.init_process_group(backend="gloo")  # type: ignore
+        return core.DistributedContext.from_torch_distributed()
+    elif info and (len(info.container_addrs) > 1 or len(info.slot_ids) > 1):
+        raise ValueError(
+            "In multi-slot managed cluster training, you must wrap your training script with a "
+            "distributed launch layer such as determined.launch.torch_distributed or "
+            "determined.launch.horovod"
+        )
+    return None
+
+
+def _set_random_seeds(seed: int) -> None:
+    # Set identical random seeds on all training processes.
+    # When using horovod, each worker will start at a unique
+    # offset in the dataset, ensuring that it is processing a unique
+    # training batch.
+    random.seed(seed)
+    np.random.seed(seed)
+    torch.random.manual_seed(seed)
+
+
+def _generate_local_seed() -> int:
+    return random.randint(0, 1 << 31)
+
+
+@contextlib.contextmanager
+def init(
+    *,
+    hparams: Optional[Dict] = None,
+    exp_conf: Optional[Dict[str, Any]] = None,
+    distributed: Optional[core.DistributedContext] = None
+) -> pytorch.PyTorchTrialContext:
+    cluster_info = det.get_cluster_info()
+    local_training = cluster_info is None or cluster_info.task_type != "TRIAL"
+
+    # Pre-execute steps: initialize distributed backend and random seeds
+    distributed_context = distributed
+
+    if not local_training:
+        distributed_context = _initialize_distributed_backend()
+
+    # Initialize default values
+    if local_training:
+        trial_seed = _generate_local_seed()
+
+        # XXX: figure out if better way to handle this
+        aggregation_frequency = exp_conf and int(exp_conf.get("optimizations", {}).get("aggregation_frequency", 1)) or 1  # type: ignore
+        fp16_compression = False
+        average_aggregated_gradients = True
```

Comment:
these backwards-compatibility things also need to find their way out of `det.pytorch.init()`.